source/grid/grid_tools.cc

```diff
@@ -2870,6 +2870,180 @@ namespace GridTools
+         coarse_cell != local_triangulation.end(); ++coarse_cell, ++index)
+      {
+        patch_to_global_tria_map_tmp.insert (std::make_pair(coarse_cell, uniform_cells[index]));
+        // To ensure that the cells with the same coordinates (here, we compare their centers) are mapped into each other.
+
+        Assert(coarse_cell->center().distance( uniform_cells[index]->center())<=1e-15*coarse_cell->diameter(),
+               ExcInternalError());
+      }
+    bool refinement_necessary;
+    // In this loop we start to do refinement on the above coarse triangulation to reach
+    // to the same level of refinement as the patch cells are really on
+    do
+      {
+        refinement_necessary = false;
+        for (typename Triangulation<Container::dimension,Container::space_dimension>::active_cell_iterator
+             active_tria_cell = local_triangulation.begin_active();
+             active_tria_cell != local_triangulation.end(); ++active_tria_cell)
+          {
+            if (patch_to_global_tria_map_tmp[active_tria_cell]->has_children())
+              {
+                active_tria_cell -> set_refine_flag();
+                refinement_necessary = true;
+              }
+            else for (unsigned int i=0; i<patch.size(); ++i)
+                {
+                  if (patch_to_global_tria_map_tmp[active_tria_cell]==patch[i])
+                    {
+                      active_tria_cell->set_user_flag();
+                      break;
+                    }
+                }
+          }
+
+        if (refinement_necessary)
+          {
+            local_triangulation.execute_coarsening_and_refinement ();
+
+            for (typename Triangulation<Container::dimension,Container::space_dimension>::cell_iterator
+                 cell = local_triangulation.begin();
+                 cell != local_triangulation.end(); ++cell)
+              {
+
+                if (patch_to_global_tria_map_tmp.find(cell)!=patch_to_global_tria_map_tmp.end())
+                  {
+                    if (cell-> has_children())
+                      {
+                        // Note: Since the cell got children, then it should not be in the map anymore
+                        // children may be added into the map, instead
+
+                        // these children may not yet be in the map
+                        for (unsigned int c=0; c< cell ->n_children(); ++c)
+                          {
+                            if (patch_to_global_tria_map_tmp.find(cell->child(c)) ==
+                                patch_to_global_tria_map_tmp.end())
+                              {
+                                patch_to_global_tria_map_tmp.insert (std::make_pair(cell ->child(c),
+                                                                                    patch_to_global_tria_map_tmp[cell]->child(c)));
+
+                                Assert(cell->child(c)->center().distance( patch_to_global_tria_map_tmp[cell]->child(c)->center())
```

Comment:
On 01/18/2016 03:24 PM, Spencer Patty wrote:
> If we happen to have a grid that has had the vertices perturbed, for instance
> using the GridTools::distort_random() member function (which is useful for
> removing symmetries of a Grid for testing an algorithm convergence rate) The
> triangulation we construct by going back to the coarsest common level of the
> patch and then refining is in fact a different triangulation than the one we
> are trying to match. In particular, the assertions here of cell centers being
> the same is triggered. What this means is that for this implementation, and
> invariant is that the triangulation needs to have been constructed under the
> exact same rules as the dealii refinement strategy of using midpoints. (This
> might be an issue say for an imported triangulation or one that has been
> smoothed, or one that has applied the random perturbation of vertices, and
> maybe others...)
>

That's a good point. I think the right way to address this is to, at the end 
of the process where we have built the map between cells on the two 
triangulation, to have a loop of the form`
```
   for (patch_cell=...)
     for (v=0...)
       patch_cell->vertex(v) = cell_map[patch_cell]->vertex(v)
```

My preference would be that we get the current patch into a shape so that it 
can be merged, and make this modification in a follow-up patch (accompanied by 
an appropriate testcase). Would that work? Care to open an "issue" on github 
with such a course of action so we don't forget?